# HG changeset patch
# User Rik <rik@octave.org>
# Date 1553714705 25200
#      Wed Mar 27 12:25:05 2019 -0700
# Branch stable
# Node ID a7ad08ed3cb75b6c816e844af9ce970cc5da22f4
# Parent  1e5dde93eb26d83457c45dbfa75b8bc2551aa266
waitbar.m: Fix hang when using "createcancelbtn" property (bug #55963).

* waitbar.m: Don't use addlistener on "createcancelbtn" property until after
properties have been applied to waitbar figure.  Tweak code to conform to
Octave style guidelines.

diff --git a/scripts/gui/waitbar.m b/scripts/gui/waitbar.m
--- a/scripts/gui/waitbar.m
+++ b/scripts/gui/waitbar.m
@@ -127,26 +127,34 @@ function h = waitbar (varargin)
 
     ax = axes ("parent", hf,
                "xtick", [], "ytick", [],
                "xlim", [0, 1], "ylim", [0, 1],
                "position", [0.1, 0.3, 0.8, 0.2]);
 
     ## Add createcancelbtn property
     addproperty ("createcancelbtn", hf, "figurebuttondownfcn");
-    addlistener (hf, "createcancelbtn", {@updatecancelbutton, ax});
+    ## FIXME: Can't add listener because of bug #55963.
+    #addlistener (hf, "createcancelbtn", {@updatecancelbutton, ax});
 
     if (! isempty (varargin))
       set (hf, varargin{:});
     endif
 
+    ## Add listener and create cancel button only after setting properties
+    ## which could change "createcancelbtn" property (bug #55963).
+    addlistener (hf, "createcancelbtn", {@updatecancelbutton, ax});
+    if (any (strcmp ("createcancelbtn", varargin)))
+      updatecancelbutton (hf, [], ax);
+    endif
+
     hp = patch (ax, [0; frac; frac; 0], [0; 0; 1; 1], [0, 0.35, 0.75]);
 
     ## Cache the axes and patch handles.
-    set (hf, "__guidata__", [ax hp]);
+    set (hf, "__guidata__", [ax, hp]);
 
     if (! (ischar (msg) || iscellstr (msg)))
       msg = "Please wait...";
     endif
     title (ax, msg);
 
     set (0, "currentfigure", cf);
   endif
@@ -157,53 +165,53 @@ function h = waitbar (varargin)
     h = hf;
   endif
 
   ## If there were no errors, update current waitbar.
   curr_waitbar = hf;
 
 endfunction
 
-function updatecancelbutton (hf, dummy, hax)
+function updatecancelbutton (hf, ~, hax)
 
-  if (! strcmpi (get (hf, "__graphics_toolkit__"), "qt"))
-    return
+  if (! strcmp (get (hf, "__graphics_toolkit__"), "qt"))
+    return;
   endif
 
   hbtn = findobj (hf, "type", "uicontrol", "-and", "style", "pushbutton");
   cb = get (hf, "createcancelbtn");
   if (! isempty (cb))
     if (isempty (hbtn))
       units = get (hax, "units");
       fpos = get (hf, "position");
       set (hax, "units", "pixels");
       apos = get (hax, "position");
 
-      fpos (2) -= 40;
-      fpos (4) += 40;
-      apos (2) += 40;
+      fpos(2) -= 40;
+      fpos(4) += 40;
+      apos(2) += 40;
       set (hf, "position", fpos);
       set (hax, "position", apos, "units", units);
 
-      hbtn = uicontrol ("style", "pushbutton", "string", "Cancel", ...
-                        "position", [fpos(3)-100 10 60 25],...
-                        "callback", cb, "parent", hf);
+      hbtn = uicontrol (hf, "style", "pushbutton", "string", "Cancel", ...
+                            "position", [fpos(3)-100, 10, 60, 25], ...
+                            "callback", cb);
     else
       set (hbtn, "callback", cb);
     endif
   elseif (! isempty (hbtn))
     delete (hbtn);
     units = get (hax, "units");
     fpos = get (hf, "position");
     set (hax, "units", "pixels");
     apos = get (hax, "position");
 
-    fpos (2) += 40;
-    fpos (4) -= 40;
-    apos (2) -= 40;
+    fpos(2) += 40;
+    fpos(4) -= 40;
+    apos(2) -= 40;
     set (hf, "position", fpos);
     set (hax, "position", apos, "units", units);
   endif
 
 endfunction
 
 
 %!demo
